dev/run-tests.py

```diff
@@ -0,0 +1,506 @@
+
+    rm_r("lib_managed")
+
+    if build_tool == "maven":
+        build_spark_maven(hadoop_version)
+    else:
+        build_spark_sbt(hadoop_version)
+
+
+def detect_binary_inop_with_mima():
+    set_title_and_block("Detecting binary incompatibilities with MiMa",
+                        "BLOCK_MIMA")
+    run_cmd(["./dev/mima"])
+
+
+def determine_test_modules(test_env):
+    """This function current acts to determine if SQL tests need to be run in
+    addition to the core test suite *or* if _only_ SQL tests need to be run
+    as the git logs show that to be the only thing touched. In the future
+    this function will act more generically to help further segregate the
+    test suite runner (hence the function name).
+    @return a set of unique test names"""
+    test_suite = list()
+
+    if test_env == "amplab_jenkins":
+        target_branch = os.environ.get("ghprbTargetBranch")
+
+        run_cmd(['git', 'fetch', 'origin', str(target_branch+':'+target_branch)])
+
+        raw_output = subprocess.check_output(['git', 'diff', '--name-only', target_branch])
+        # remove any empty strings
+        changed_files = [f for f in raw_output.split('\n') if f]
+
+        # find any sql files
+        sql_files = [f for f in changed_files
+                     if any(f.startswith(p) for p in
+                            ["sql/",
+                             "bin/spark-sql",
+                             "sbin/start-thriftserver.sh",
+                             "examples/src/main/java/org/apache/spark/examples/sql/",
+                             "examples/src/main/scala/org/apache/spark/examples/sql/"])]
+        mllib_files = [f for f in changed_files
+                       if any(f.startswith(p) for p in
+                              ["examples/src/main/java/org/apache/spark/examples/mllib/",
+                               "examples/src/main/scala/org/apache/spark/examples/mllib",
+                               "data/mllib/",
+                               "mllib/"])]
+        streaming_files = [f for f in changed_files
+                           if any(f.startswith(p) for p in
+                                  ["examples/scala-2.10/",
+                                   "examples/src/main/java/org/apache/spark/examples/streaming/",
+                                   "examples/src/main/scala/org/apache/spark/examples/streaming/",
+                                   "external/",
+                                   "extras/java8-tests/",
+                                   "extras/kinesis-asl/",
+                                   "streaming/"])]
+        graphx_files = [f for f in changed_files
+                        if any(f.startswith(p) for p in
+                               ["examples/src/main/scala/org/apache/spark/examples/graphx/",
+                                "graphx/"])]
+
+        non_sql_files = set(changed_files).difference(set(sql_files))
+
+        if non_sql_files:
```

Comment:
The logic here seems to be "if any file outside of SQL changed, then re-run the core tests", but I think we really want something more like "if any file that's not part of one of the high-level libraries has changed, then re-run core."  I think that the current logic will end up retesting all of core whenever streaming changes, for example.

I think that an easy short-term fix for this would be to change this code so that we define a new set named `non_high_level_library_files` (or maybe something less verbose) and compute the difference between `changed_files` and all of the high-level library files.  I considered maybe naming this `core_files`, but that's not quite the opposite of SQL files, since it also includes high-level libraries that aren't special-cased here (like Bagel or the Kafka streaming tests) as well as changes that only affect project infra (such as build scripts).